library(RUnit)
source("./tmsRunner.R")
trenaProject <- TrenaProjectErythropoiesis()
tbl.atac.merged <- get(load("~/github/TrenaProjectErythropoiesis/inst/extdata/genomicRegions/tbl.atacMerged.RData"))


x <- TMS$new("CHD1", trenaProject)
x$addGeneHancer()
tbl.gh <- x$getGeneHancer()
#checkEquals(dim(tbl.gh), c(25, 16))
tbl.ghRegion <- x$getGeneHancerRegion()
#checkEquals(tbl.ghRegion$start, min(tbl.gh$start) - 1000)
#checkEquals(tbl.ghRegion$end, max(tbl.gh$end) + 1000)
x$addOpenChromatin()
tbl.oc <- x$getOpenChromatin()
browser()
x$viewOpenChromatin()
#checkEquals(dim(tbl.oc), c(88, 7))
motifs <- query(MotifDb, c("sapiens"), c("jaspar2018", "hocomocov11-core"))
length(motifs)
x$addAndScoreFimoTFBS(fimo.threshold=1e-6, motifs=motifs)
tbl.tms <- x$getTfTable()
#checkEquals(dim(tbl.tms), c(48, 16))
x$addRBP()
tbl.rbp <- x$getRbpTable()
getExpressionMatrixNames(trenaProject)
mtx <- getExpressionMatrix(trenaProject,
                           "brandLabDifferentiationTimeCourse-27171x28-namesCorrected")
x$add.tf.mrna.correlations(mtx, featureName="cor.all")
tbl.tms <- x$getTfTable()
#checkTrue("cor.all" %in% colnames(tbl.tms))
fivenum(tbl.tms$cor.all)
x$add.rbp.mrna.correlations(mtx, featureName="cor.all")
tbl.rbp <- x$getRbpTable()
#checkTrue("cor.all" %in% colnames(tbl.rbp))
fivenum(tbl.rbp$cor.all)
dim(tbl.rbp)

tfs <- unique(subset(tbl.tms, gh > 600 & abs(cor.all) > 0.7)$tf)
length(tfs)
rbps <- unique(subset(tbl.rbp, celltype=="K562" & abs(cor.all) > 0.7)$gene)
length(rbps)
tbl.trena <- x$build.trena.model(tfs, rbps, mtx)

model <- x$build.linear.model(sort.by.column="betaLasso")
